modules/codegen-plugin/src/smithy4s/codegen/Smithy4sCodegenPlugin.scala

```diff
@@ -403,8 +403,8 @@ object Smithy4sCodegenPlugin extends AutoPlugin {
       (inputDirs ++ generatedFiles)
         .filter(_.exists())
         .toList
```

Comment:
**question:** what about Mill? I think it has the same problem:

```scala
def smithy4sOutputDir: T[PathRef] = T {
  PathRef(T.ctx().dest / "scala")
}

def smithy4sResourceOutputDir: T[PathRef] = T {
  PathRef(T.ctx().dest / "resources")
}
```

cached `PathRef` state will be invalidated if the contents change. It should probably be just `os.Path` in these settings (this might have to wait for 0.19 because we'd change public type signatures).

Also, even today this probably isn't idiomatic Mill: `T{...}` doesn't get recomputed, but it's wrapping a mutable input here (`PathRef`). Should probably be `T.input`, but then you'd see the problem from #1495 quite prominently in Mill.

Reply:
Nah, mill is fine. For starters, downstream tasks need to know when their upstreams change, that's what the pathref is for. 

Secondly, the dest call means that the location is specific to the task, which implies a separation between different plugins, unlike what we had in SBT

Reply:
> Secondly, the dest call means that the location is specific to the task, which implies a separation between different plugins, unlike what we had in SBT

I agree that it does have the separation, it's not what I'm arguing.

> downstream tasks need to know when their upstreams change, that's what the pathref is for.

yeah, and I'm saying that changes to the upstream (`smithy4sOutputDir`) don't invalidate the downstream _because_ the output dir is a Target:

```scala
import $ivy.`com.disneystreaming.smithy4s::smithy4s-mill-codegen-plugin::0.18.16`

import smithy4s.codegen.mill._
import mill._
import mill.scalalib._

object service extends Smithy4sModule {

  def millSourcePath = os.pwd

  def scalaVersion = "3.3.3"

  def foo = T {
    T.ctx().log.info(s"recomputing foo at ${java.time.Instant.now()}!")
    os.walk(smithy4sOutputDir().path).size
  }

}
```

demo: the actual file count changes, but not the one `foo` sees (because `foo` doesn't re-run, because its input, `smithy4sOutputDir`, didn't change, because it memoizes the path):

<img width="1339" alt="image" src="https://github.com/disneystreaming/smithy4s/assets/894884/3e907b95-45ab-4c33-b66e-1342711d68d3">

so to sum up, I think the Mill plugin is in the limbo of not deciding whether regenerating code: should result in invalidating the cache of the output path (`PathRef` would do that), or not (`T{...}` doesn't invalidate).

For what it's worth, I think just slapping on `T.input` will technically help tasks like `foo`, but it will probably also break the plugin because the output path, being an input to the codegen, will invalidate the codegen task.

A proper solution in my mind looks more like this:

1. turn `smithy4sOutputDir` into an `os.Path`, mirroring what Michał is doing in this PR
2. anyone who wishes to be notified of its changes will wrap it in a `PathRef` in their own `T.input`

so:

```scala
//smithy4s-mill

def smithy4sOutputDir: T[os.Path] = T { ... }

// use-site of whatever wants to be notified of the output dir's current state
def smithy4sOutputPath: T[PathRef] = T.input { PathRef(smithy4sOutputDir()) }

// example task using the thing
def foo = T {
    //this now gets re-run every time generated sources are changed
    os.walk(smithy4sOutputPath().path).size
}

```

Reply:
Oh, you're right.  It'd be worth asking the mill folks what they think.

Like, ideally, the codegen task would only every be able to produce code in its own dest, sandbox style, and you'd be able to inspect the path to the dest folder of a task statically without actually running it. Doing that is really important in the context of scala LSP, where you shouldn't have to run anything to know the source directories of a module.

Anyway, that's the rationale for having the smithy4s output path as a separate task : we shouldn't have to run the codegen to know where it outputs things. 

But you are correct, the PathRef-returning task should be an input rather than a target.

Reply:
moved to new ticket: https://github.com/disneystreaming/smithy4s/issues/1529